src/librustc/traits/select.rs

```diff
@@ -78,8 +77,40 @@ pub struct SelectionContext<'cx, 'tcx:'cx> {
     /// there is no type that the user could *actually name* that
     /// would satisfy it. This avoids crippling inference, basically.
     intercrate: bool,
+
+    /// Obligations generated during inference. If None, no obligations are to be collected (e.g.
+    /// this is swapped out to a `None` when probing).
+    inferred_obligations: Option<PredicateObligations<'tcx>>
```

Comment:
Hmm. I might prefer `SnapshotVec`, yeah, coupled with some assertions
that the vector is empty each time we begin an operation. One thing that
worries me in the current setup is that we have `None` by default, so if
we *forget* to swap in a vec, things get dropped.

On Fri, May 06, 2016 at 03:11:31PM -0700, Masood Malekghassemi wrote:
> > @@ -78,8 +77,40 @@ pub struct SelectionContext<'cx, 'tcx:'cx> {
> >      /// there is no type that the user could *actually name* that
> >      /// would satisfy it. This avoids crippling inference, basically.
> >      intercrate: bool,
> > +
> > +    /// Obligations generated during inference. If None, no obligations are to be collected (e.g.
> > +    /// this is swapped out to a `None` when probing).
> > +    inferred_obligations: Option<PredicateObligations<'tcx>>
> 
> What about using `SnapshotVec`? It's roughly the same principle (even using fewer allocs, perhaps), but: would that alleviate concerns w.r.t. the swapping?
> 
> ---
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub:
> https://github.com/rust-lang/rust/pull/33301/files/b7caae30c7389c3b92a8399577e007116c28f075..bf98976ec559d36f2aac16c1552155f417046afc#r62397872